common/enum_mask_base.h

```diff
@@ -0,0 +1,154 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_ENUM_MASK_BASE_H_
+#define CARBON_COMMON_ENUM_MASK_BASE_H_
+
+#include <bit>
+
+#include "common/enum_base.h"
+#include "llvm/ADT/StringExtras.h"
+
+namespace Carbon::Internal {
+
+// CRTP-style base class similar to `EnumBase`, but supporting mask enums.
+//
+// Users must be in the `Carbon` namespace and should look like the following.
+//
+// In `my_kind.h`:
+//   ```
+//   #define CARBON_MY_KIND(X) \
+//       X(Enumerator1)        \
+//       X(Enumerator2)        \
+//       X(Enumerator3)        \
+//       ...
+//
+//   CARBON_DEFINE_RAW_ENUM_MASK(MyKind) {
+//     CARBON_MY_KIND(CARBON_RAW_ENUM_MASK_ENUMERATOR)
+//   };
+//
+//   class MyKind : public CARBON_ENUM_MASK_BASE(MyKind) {
+//    public:
+//     CARBON_MY_KIND(CARBON_ENUM_MASK_CONSTANT_DECL)
+//
+//     // Plus, anything else you wish to include.
+//   };
+//
+//   #define CARBON_MY_KIND_WITH_TYPE(X) \
+//     CARBON_ENUM_MASK_CONSTANT_DEFINITION(MyKind, X)
+//   CARBON_MY_KIND(CARBON_MY_KIND_WITH_TYPE)
+//   #undef CARBON_MY_KIND_WITH_TYPE
+//   ```
+//
+// In `my_kind.cpp`:
+//   ```
+//   CARBON_DEFINE_ENUM_MASK_NAMES(MyKind) = {
+//       CARBON_MY_KIND(CARBON_ENUM_MASK_NAME_STRING)};
```

Comment:
Somewhat unfortunate if `clang-format` won't render this as:

```suggestion
//       CARBON_MY_KIND(CARBON_ENUM_MASK_NAME_STRING)
//    };
```

But super minor. Mostly mentioning in case this was manual formatting -- I'd find the close curly easier to see this way given that it's always just "one" entry (that expands).

Reply:
This is automatic, I wasn't seeing an easy way to make it autoformat that way.

For context, we currently have things like:

```
CARBON_DEFINE_ENUM_CLASS_NAMES(Kind) = {
#define CARBON_DIAGNOSTIC_KIND(Name) CARBON_ENUM_CLASS_NAME_STRING(Name)
#include "third_party/carbon/lang/toolchain/diagnostics/diagnostic_kind.def"
};
```

Here, the alternate syntax means we don't notice the formatting consequence. The `=` is significant, though: previously the decision had been to include it because this is defining a variable, versus the other case that's defining enum contents. If we were to drop the `=`, then `IfMacros` could be applied and should end up with the wrapping behavior you want, but not necessarily the overall formatting.

Reply:
See #6054